File: api/clipcatalyst_api/brandkit.py
"""Brand kit: the cloud half of the shared core (BRANDKIT.md, Module 3).

A direct port of the pure section of ``lib/studio/brandkit.ts`` — the overlay
geometry, the colour rule both engines obey, and the upload limits — plus the
two things only a server can do: sniff the bytes it was actually handed
instead of trusting a header, and derive a storage path from the account id
that cannot leave the brand directory whatever the upload called itself.

Two rules carry over from the TypeScript verbatim, and everything else follows
from them:

  1. A brand kit must NEVER be the reason a render fails. Every function here
     is total — garbage in gets a truthful fallback out, never a raise.
     ``normalize_hex`` answers None, ``logo_box`` answers a usable box, an
     unreadable logo answers "no overlay" back in render.py.
  2. Diarization wins over the brand colour. With two or more speakers the
     palette keeps colouring the active word, because that colour says WHO is
     talking; the brand colour replaces the default violet for unassigned
     words and single-speaker clips only. ``active_word_color`` is the ONE
     place that decides it, exactly as ``activeWordColor`` is over there.

The plan gate lives with the callers (main.py for storage, worker.py for
render time), not here: this module knows nothing about plans, and
``watermark_required`` precedence — our mark beats their logo — is the
renderer's to apply.

``logo_box`` is cross-checked against the TypeScript by a node subprocess in
``api/tests/test_brandkit.py``, like croptrack and diarize, so a drift in
either direction fails a test rather than shipping two different corners.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from pathlib import Path

from .pipeline.diarize import SPEAKER_COLORS
from .settings import Settings

# --- Tunables. The same numbers as lib/studio/brandkit.ts, pinned there by
#     scripts/brandkit.test.mjs and here by api/tests/test_brandkit.py. -------

#: Upload ceiling, applied to the bytes RECEIVED (never to a declared length).
MAX_LOGO_BYTES = 2_000_000
#: Accepted upload types. SVG is allowed: the browser rasterizes it via <img>,
#: ffmpeg via its own decoder — and when a build has no SVG decoder the render
#: simply carries no logo instead of failing (see render._logo_box_for).
LOGO_TYPES: tuple[str, ...] = (
    "image/png",
    "image/jpeg",
    "image/webp",
    "image/svg+xml",
)
#: The violet the caption highlight has always used (== SPEAKER_COLORS[0]).
DEFAULT_CAPTION_COLOR = "#a78bfa"
#: Logo height as a fraction of output height.
LOGO_HEIGHT_RATIO = 0.045
#: Width ceiling as a fraction of output width, for wordmark-shaped logos.
LOGO_MAX_WIDTH_RATIO = 0.32
#: Corner margin, as a fraction of output HEIGHT on both axes.
LOGO_MARGIN_RATIO = 0.02

#: The ONLY filename extensions a stored logo can have: one per accepted type,
#: chosen by the SNIFFED type. Nothing derived from the upload's own filename
#: ever reaches the filesystem (see logo_storage_name).
LOGO_EXTENSIONS: dict[str, str] = {
    "image/png": "png",
    "image/jpeg": "jpg",
    "image/webp": "webp",
    "image/svg+xml": "svg",
}

#: The inverse map, for serving a stored logo back with an honest type. Built
#: from LOGO_EXTENSIONS so the two can never drift apart.
LOGO_MEDIA_TYPES: dict[str, str] = {
    extension: content_type for content_type, extension in LOGO_EXTENSIONS.items()
}

#: The name of ``settings.brand_dir`` inside the data dir. A logo is stored on
#: the user row as "<BRAND_DIR_NAME>/<user id>.<ext>" — relative, so the row
#: stays portable across boxes whose data dirs differ.
BRAND_DIR_NAME = "brand"

#: An account id is a uuid4 hex, but founder-edited rows exist; the filename
#: is built only from ids that are already safe, rather than by sanitizing a
#: hostile one into looking safe.
_SAFE_USER_ID_RE = re.compile(r"^[A-Za-z0-9_-]{1,64}$")

_PNG_MAGIC = b"\x89PNG\r\n\x1a\n"
_JPEG_MAGIC = b"\xff\xd8\xff"
_UTF8_BOM = b"\xef\xbb\xbf"
#: How far into a file we look for an ``<svg`` root. An XML declaration plus a
#: DOCTYPE is ~150 bytes; a kilobyte is generous without being a scan.
_SVG_SNIFF_BYTES = 1024

WRONG_TYPE_MESSAGE = (
    "That file isn't an image we can use — pick a PNG, JPEG, WebP or SVG."
)
EMPTY_MESSAGE = "That file is empty — pick a PNG, JPEG, WebP or SVG."


@dataclass(frozen=True)
class LogoBox:
    """Pixel box the logo is drawn into, top-left origin, whole pixels."""

    x: int
    y: int
    width: int
    height: int


def _js_round(x: float) -> int:
    """``Math.round`` semantics: nearest integer, ties toward +infinity.

    Python's built-in ``round`` is banker's rounding, so it answers 22 where
    JavaScript answers 23 for 22.5 — and a 480-tall export puts the logo
    height on exactly that half. Same helper, same reason, as diarize's
    ``_js_round``; the parity test would catch its absence at 500 and 1500.
    """
    if not math.isfinite(x):
        return 0
    floor = math.floor(x)
    return floor + 1 if x - floor >= 0.5 else floor


def _finite_or(value: float | None, fallback: float) -> float:
    """``finiteOr`` from the TypeScript: NaN/inf/None read as the fallback."""
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return float(value) if math.isfinite(value) else fallback
    return fallback


# --- Colour ----------------------------------------------------------------


def normalize_hex(value: str | None) -> str | None:
    """``"#abc"`` → ``"#aabbcc"``, ``"#AABBCC"`` → ``"#aabbcc"``; else None.

    The same rule as ``normalizeHex`` in the TypeScript, so a colour that
    survives the panel survives the upload: surrounding whitespace is forgiven
    (a human types into that field), nothing else is — no missing hash, no
    named colours, no ``rgb()``, no 4/8-digit alpha forms.
    """
    if not isinstance(value, str):
        return None
    raw = value.strip().lower()
    if len(raw) not in (4, 7) or not raw.startswith("#"):
        return None
    body = raw[1:]
    if any(ch not in "0123456789abcdef" for ch in body):
        return None
    if len(body) == 3:
        return f"#{body[0] * 2}{body[1] * 2}{body[2] * 2}"
    return f"#{body}"


def active_word_color(
    speaker: int | None, speaker_count: int, caption_color: str | None
) -> str:
    """The colour one active caption word takes — mirrors ``activeWordColor``.

    Diarization wins: with 2+ speakers an assigned word keeps its speaker
    colour. Everything else — unassigned words, single-speaker clips — takes
    the brand colour when there is one, else the default violet, which IS
    ``SPEAKER_COLORS[0]``, so an unbranded clip is coloured exactly as it was
    before brand kits existed. A stored colour that no longer parses degrades
    to the default rather than emitting a broken value into an ASS file.
    """
    if speaker is not None and speaker_count >= 2:
        return SPEAKER_COLORS[int(speaker) % len(SPEAKER_COLORS)]
    return normalize_hex(caption_color) or DEFAULT_CAPTION_COLOR


# --- Geometry --------------------------------------------------------------


def logo_box(
    natural_width: float, natural_height: float, out_width: float, out_height: float
) -> LogoBox:
    """The box the logo is drawn into, bottom-right, aspect preserved.

    Height is ``LOGO_HEIGHT_RATIO`` of the output height, width follows the
    natural aspect, and a wordmark too wide for ``LOGO_MAX_WIDTH_RATIO`` of the
    output width is clamped with the height RE-DERIVED so it never stretches.
    The margin is ``LOGO_MARGIN_RATIO`` of the output height on both axes.

    Line for line the same as ``logoBox`` in lib/studio/brandkit.ts, including
    the rounding: whole pixels out, because a half-pixel overlay origin makes
    ffmpeg resample the logo. Total by construction — a zero, negative or
    non-finite natural size reads as a square rather than a division by zero,
    and the box is never smaller than 1x1.
    """
    out_w = max(1, _js_round(_finite_or(out_width, 0)))
    out_h = max(1, _js_round(_finite_or(out_height, 0)))
    nat_w = _finite_or(natural_width, 0)
    nat_h = _finite_or(natural_height, 0)
    # Unknown natural size (an SVG with no intrinsic dimensions, a logo that
    # failed to decode) reads as square rather than as a divide by zero.
    aspect = nat_w / nat_h if nat_w > 0 and nat_h > 0 else 1.0

    height = out_h * LOGO_HEIGHT_RATIO
    width = height * aspect
    max_width = out_w * LOGO_MAX_WIDTH_RATIO
    if width > max_width:
        width = max_width
        height = width / aspect
    w = max(1, _js_round(width))
    h = max(1, _js_round(height))
    margin = _js_round(out_h * LOGO_MARGIN_RATIO)
    # The origin only needs clamping for absurd frames (taller than they are
    # wide by more than ~16:1), where the height-derived margin outgrows the
    # width. An off-frame overlay origin would be an ffmpeg error rather than
    # a missing logo, so it is guarded.
    return LogoBox(
        x=max(0, out_w - margin - w),
        y=max(0, out_h - margin - h),
        width=w,
        height=h,
    )


# --- Uploads ---------------------------------------------------------------


def format_mb(size: int) -> str:
    """Megabytes for a message, rounded UP to the nearest 10 KB.

    ``formatMb``'s rule, so a file one byte over the limit never reads as
    "That logo is 2 MB — the limit is 2 MB", and the server's refusal is
    worded identically to the one the panel would have shown — including the
    number's spelling: JavaScript prints a whole 2 as "2", not "2.0".
    """
    value = math.ceil(size / 10_000) / 100
    return str(int(value)) if float(value).is_integer() else str(value)


def sniff_image_type(data: bytes) -> str | None:
    """The image type these BYTES actually are, or None if not one we accept.

    The upload's own ``Content-Type`` is not consulted anywhere: it is client
    input, and the whole point of a whitelisted extension is undone if the
    client picks which one applies. Magic numbers for the raster formats; a
    real ``<svg`` root element for SVG.
    """
    if not data:
        return None
    if data.startswith(_PNG_MAGIC):
        return "image/png"
    if data.startswith(_JPEG_MAGIC):
        return "image/jpeg"
    if len(data) >= 12 and data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        return "image/webp"
    if _looks_like_svg(data):
        return "image/svg+xml"
    return None


def _looks_like_svg(data: bytes) -> bool:
    """An XML document whose markup opens an ``<svg`` root, near the top."""
    head = data[:_SVG_SNIFF_BYTES]
    if head.startswith(_UTF8_BOM):
        head = head[len(_UTF8_BOM) :]
    head = head.lstrip()
    return head.startswith(b"<") and b"<svg" in head.lower()


def validate_logo_bytes(data: bytes) -> tuple[str | None, str | None]:
    """``(sniffed content type, error message)`` — exactly one is not None.

    The messages are the ones ``validateLogoFile`` shows, verbatim, because
    they are shown to the creator: they say what is wrong and what the limit
    is, never "invalid file".
    """
    if not data:
        return None, EMPTY_MESSAGE
    if len(data) > MAX_LOGO_BYTES:
        return None, (
            f"That logo is {format_mb(len(data))} MB — the limit is "
            f"{format_mb(MAX_LOGO_BYTES)} MB."
        )
    content_type = sniff_image_type(data)
    if content_type is None:
        return None, WRONG_TYPE_MESSAGE
    return content_type, None


# --- Storage paths ---------------------------------------------------------


def logo_storage_name(user_id: str, content_type: str) -> str | None:
    """``"brand/<user id>.<ext>"``, or None when the id has no safe form.

    Path safety is structural rather than a filter: the leaf is the account id
    (already a uuid4 hex) plus an extension chosen from LOGO_EXTENSIONS by the
    sniffed type. A ``../../etc/passwd`` filename on the upload cannot reach
    this function's output because the upload's filename is never an input.
    """
    extension = LOGO_EXTENSIONS.get(content_type)
    if extension is None or not _SAFE_USER_ID_RE.match(user_id or ""):
        return None
    return f"{BRAND_DIR_NAME}/{user_id}.{extension}"


def logo_abs_path(settings: Settings, stored: str) -> Path | None:
    """A stored logo name resolved to a real path inside the brand dir.

    None when the row is empty, or when it resolves anywhere else — the row is
    written by ``logo_storage_name`` alone, so this only bites on a hand-edited
    database, which is exactly the case worth refusing to read a file for.
    """
    if not stored:
        return None
    root = settings.brand_dir
    resolved = (settings.data_dir / stored).resolve()
    try:
        inside = resolved.is_relative_to(root.resolve())
    except (OSError, ValueError):
        return None
    return resolved if inside else None

File: lib/studio/brandkit.ts
// Brand kit: a creator's own logo in the corner and their own colour on the
// caption highlight. This module is the shared core BOTH renderers obey — the
// browser canvas path (render.ts) and the ffmpeg/ASS path (api/pipeline) — so
// the geometry and the colour rule live here exactly once and the Python side
// is a direct port with a cross-check test, like croptrack and diarize.
//
// Two rules are worth stating up front because every other decision follows:
//
//   1. A brand kit must NEVER be the reason a render fails. Every function
//      below is total: garbage in gets a truthful fallback out, not a throw.
//      normalizeHex answers null, logoBox answers a usable box, loadKit
//      answers EMPTY_KIT. The corner drawing is a nicety; the clip is not.
//   2. Diarization wins over the brand colour. When a clip has two or more
//      speakers, SPEAKER_COLORS keeps colouring the active word — that colour
//      carries information (who is talking) the brand colour would destroy.
//      The brand colour replaces the default violet for unassigned words and
//      single-speaker clips only. activeWordColor is the ONE place that
//      decides this, and both engines call through it.
//
// The plan gate lives with the caller, not here: `watermark_required` on the
// plan means our mark is drawn and the kit's logo is not, and that precedence
// is the renderer's to apply. This module knows nothing about plans.
//
// Everything above the IO section is pure and has ZERO runtime imports (the
// SPEAKER_COLORS mirror below is deliberate) — the compiled JS is loaded
// straight out of .croptrack-build by the node harness.

/** The whole feature. No fonts, no caption templates, no secondary colours —
 *  those are separate features and pretending otherwise is how the pricing
 *  page got into trouble in the first place. */
export type BrandKit = {
  /** Logo drawn bottom-right in place of the ClipCatalyst mark. */
  logo: { dataUrl: string; width: number; height: number } | null;
  /** Caption highlight colour, "#rrggbb". null = the default violet. */
  captionColor: string | null;
  /** Draw the logo at all. Off = a clean corner (paid plans only). */
  showLogo: boolean;
};

/** The logo half of a kit, named for callers that pass it around alone. */
export type BrandLogo = NonNullable<BrandKit["logo"]>;

/** Pixel box a logo is drawn into, top-left origin, whole pixels. */
export type LogoBox = { x: number; y: number; width: number; height: number };

// --- Tunables. Every one of these is pinned by scripts/brandkit.test.mjs. ---

/** Upload ceiling. 2 MB of data URL is already ~2.7 MB of base64 in the
 *  record, which is why logos live in IndexedDB and not localStorage. */
export const MAX_LOGO_BYTES = 2_000_000;
/** Accepted upload types. SVG is allowed: both engines rasterize it (the
 *  browser via <img>, ffmpeg via its own decoder) at the size we ask for. */
export const LOGO_TYPES: readonly string[] = [
  "image/png",
  "image/jpeg",
  "image/webp",
  "image/svg+xml",
];
/** The violet the caption highlight has always used. */
export const DEFAULT_CAPTION_COLOR = "#a78bfa"; // == SPEAKER_COLORS[0]
/** Logo height as a fraction of output height — small enough to read as a
 *  mark rather than a banner at every export size. */
export const LOGO_HEIGHT_RATIO = 0.045;
/** Width ceiling as a fraction of output width, for wordmark-shaped logos.
 *  A 4:1 logo on 9:16 lands exactly on this line (0.045 × 4 == 0.32 × 9/16),
 *  so anything wider than a 4:1 wordmark is what actually gets clamped. */
export const LOGO_MAX_WIDTH_RATIO = 0.32;
/** Corner margin, as a fraction of output HEIGHT on both axes — an equal
 *  visual inset, not an equal fraction of two different edges. */
export const LOGO_MARGIN_RATIO = 0.02;

export const EMPTY_KIT: BrandKit = Object.freeze({
  logo: null,
  captionColor: null,
  showLogo: true,
});

export function isEmptyKit(kit: BrandKit): boolean {
  return kit.logo === null && kit.captionColor === null && kit.showLogo === true;
}

/**
 * Caption palette, index = speaker. diarize.ts owns SPEAKER_COLORS — mirrored
 * here because the import must stay type-only (render.ts mirrors ZOOM_RAMP_S
 * from edits.ts for the same reason). If these ever diverge, diarize.ts wins.
 */
const SPEAKER_COLORS = ["#a78bfa", "#fbbf24", "#38bdf8", "#fb7185"] as const;

// --- Colour ----------------------------------------------------------------

/**
 * "#abc" → "#aabbcc", "#AABBCC" → "#aabbcc"; anything else is null.
 *
 * Surrounding whitespace is forgiven because the UI's hex field is a text
 * input a human types into. Nothing else is: no missing hash, no named
 * colours, no rgb(), no 4/8-digit alpha forms. The Python validator applies
 * the same rule so a colour that survives here survives the upload too.
 */
export function normalizeHex(input: string): string | null {
  if (typeof input !== "string") return null;
  const raw = input.trim().toLowerCase();
  if (raw.length !== 4 && raw.length !== 7) return null;
  if (raw[0] !== "#") return null;
  const body = raw.slice(1);
  for (const ch of body) {
    if (!((ch >= "0" && ch <= "9") || (ch >= "a" && ch <= "f"))) return null;
  }
  if (body.length === 3) return `#${body[0]}${body[0]}${body[1]}${body[1]}${body[2]}${body[2]}`;
  return `#${body}`;
}

/** sRGB channel → linear light (WCAG 2.1 relative luminance). */
function channelLuminance(byte: number): number {
  const c = byte / 255;
  return c <= 0.03928 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
}

/**
 * WCAG contrast ratio of `hex` against the caption's black box: 1 (invisible)
 * to 21 (white on black), two decimals. The UI warns below 3 — it does not
 * block, because a creator picking their own dark brand colour is making a
 * choice, not a mistake. An unparseable colour answers 1, the worst case, so
 * a broken value always trips the warning rather than passing silently.
 */
export function contrastOnBlack(hex: string): number {
  const norm = normalizeHex(hex);
  if (norm === null) return 1;
  const r = channelLuminance(parseInt(norm.slice(1, 3), 16));
  const g = channelLuminance(parseInt(norm.slice(3, 5), 16));
  const b = channelLuminance(parseInt(norm.slice(5, 7), 16));
  const luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  // Black's luminance is 0, so the ratio is (L + 0.05) / (0 + 0.05).
  return Math.round(((luminance + 0.05) / 0.05) * 100) / 100;
}

/**
 * The colour the active caption word takes — the ONE place both engines
 * agree. Diarization wins: with 2+ speakers an assigned word keeps its
 * speaker colour, because that colour says WHO is talking. Everything else —
 * unassigned words, single-speaker clips — takes the brand colour when the
 * kit has one, else the default violet (== SPEAKER_COLORS[0], so an unbranded
 * clip is coloured exactly as it was before brand kits existed).
 *
 * A stored colour that no longer parses degrades to the default rather than
 * emitting a broken colour string into a canvas or an ASS file.
 */
export function activeWordColor(
  speaker: number | undefined,
  speakerCount: number,
  kit: BrandKit
): string {
  if (speaker !== undefined && Number.isFinite(speaker) && speakerCount >= 2) {
    const idx = ((Math.trunc(speaker) % SPEAKER_COLORS.length) + SPEAKER_COLORS.length) % SPEAKER_COLORS.length;
    return SPEAKER_COLORS[idx];
  }
  const brand = kit ? normalizeHex(kit.captionColor ?? "") : null;
  return brand ?? DEFAULT_CAPTION_COLOR;
}

// --- Geometry --------------------------------------------------------------

/**
 * The box the logo is drawn into, bottom-right, aspect preserved: height is
 * LOGO_HEIGHT_RATIO of the output height, width follows the natural aspect,
 * and a wordmark too wide for LOGO_MAX_WIDTH_RATIO of the output width is
 * clamped with the height RE-DERIVED so it never stretches. The margin is
 * LOGO_MARGIN_RATIO of the output height on both axes.
 *
 * This is the geometry everything else trusts: render.ts drawImage()s into
 * it, and render.py computes the same numbers for its overlay filter. Whole
 * pixels out — a half-pixel overlay origin makes ffmpeg resample the logo.
 *
 * Total by construction: a zero, negative or non-finite natural size falls
 * back to a square, and the box is never smaller than 1×1.
 */
export function logoBox(
  natural: { width: number; height: number },
  out: { width: number; height: number }
): LogoBox {
  const outW = Math.max(1, Math.round(finiteOr(out?.width, 0)));
  const outH = Math.max(1, Math.round(finiteOr(out?.height, 0)));
  const natW = finiteOr(natural?.width, 0);
  const natH = finiteOr(natural?.height, 0);
  // Unknown natural size (an SVG with no intrinsic dimensions, a logo that
  // failed to decode) reads as square rather than as a divide by zero.
  const aspect = natW > 0 && natH > 0 ? natW / natH : 1;

  let height = outH * LOGO_HEIGHT_RATIO;
  let width = height * aspect;
  const maxWidth = outW * LOGO_MAX_WIDTH_RATIO;
  if (width > maxWidth) {
    width = maxWidth;
    height = width / aspect;
  }
  const w = Math.max(1, Math.round(width));
  const h = Math.max(1, Math.round(height));
  const margin = Math.round(outH * LOGO_MARGIN_RATIO);
  // The origin only needs clamping for absurd frames (taller than they are
  // wide by more than ~16:1), where the height-derived margin outgrows the
  // width. Real export sizes never reach it; an off-canvas overlay origin
  // would be an ffmpeg error rather than a missing logo, so it is guarded.
  return {
    x: Math.max(0, outW - margin - w),
    y: Math.max(0, outH - margin - h),
    width: w,
    height: h,
  };
}

function finiteOr(value: number | undefined, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

// --- Upload validation -----------------------------------------------------

/**
 * Error message for a rejected logo, or null when it is fine. The message is
 * shown to the creator verbatim, so it says what is wrong and what the limit
 * is — never "invalid file".
 *
 * The size here is the browser's; the server re-checks the bytes it actually
 * received and sniffs the content type rather than trusting this one.
 */
export function validateLogoFile(file: { type: string; size: number }): string | null {
  if (!file || typeof file.size !== "number" || !Number.isFinite(file.size)) {
    return "That file couldn't be read. Try picking it again.";
  }
  if (typeof file.type !== "string" || !LOGO_TYPES.includes(file.type)) {
    return "That file isn't an image we can use. Pick a PNG, JPEG, WebP or SVG.";
  }
  if (file.size <= 0) {
    return "That file is empty. Pick a PNG, JPEG, WebP or SVG.";
  }
  if (file.size > MAX_LOGO_BYTES) {
    return `That logo is ${formatMb(file.size)} MB, and the limit is ${formatMb(MAX_LOGO_BYTES)} MB.`;
  }
  return null;
}

/** Megabytes for a message, rounded UP to the nearest 10 KB so a file that is
 *  one byte over never reads as "That logo is 2 MB — the limit is 2 MB". */
function formatMb(bytes: number): string {
  return String(Math.ceil(bytes / 10_000) / 100);
}

// ===========================================================================
// IO — IndexedDB, browser only. Nothing above this line touches a global.
// ===========================================================================
//
// The Studio works with no account, so the kit is local-first: one store, one
// key, the logo data URL and the colours in the same record. A 2 MB data URL
// will not fit comfortably in localStorage, which is what rules it out.
//
// Every failure path degrades to an in-memory kit for the session — blocked
// storage in private mode, a quota rejection, a browser that has no
// IndexedDB at all. The kit still applies to this session's renders; it just
// will not survive a reload. A brand kit must never break a render, and
// "could not persist" is not a reason to lose the creator's logo mid-session.
//
// Every entry point is guarded (`typeof indexedDB === "undefined"`), so
// importing this module under node or during prerender is inert.

const DB_NAME = "clipcatalyst";
const DB_VERSION = 1;
const STORE_NAME = "brandkit";
const RECORD_KEY = "kit";
/** A blocked upgrade (another tab holding the old version) leaves open()
 *  pending forever — fall back rather than hang the panel. */
const IDB_TIMEOUT_MS = 3_000;

/** The session's kit when IndexedDB is unavailable or refused the write. */
let memoryKit: BrandKit | null = null;

function hasIndexedDB(): boolean {
  // Some private modes expose the global as null rather than removing it.
  return typeof indexedDB !== "undefined" && !!indexedDB;
}

/** Whatever came out of storage → a kit that is safe to render with. */
function coerceKit(value: unknown): BrandKit {
  if (!value || typeof value !== "object") return EMPTY_KIT;
  const raw = value as Partial<BrandKit>;
  const logo = raw.logo;
  const okLogo =
    !!logo &&
    typeof logo === "object" &&
    typeof logo.dataUrl === "string" &&
    logo.dataUrl.startsWith("data:") &&
    Number.isFinite(logo.width) &&
    Number.isFinite(logo.height);
  return {
    logo: okLogo ? { dataUrl: logo.dataUrl, width: logo.width, height: logo.height } : null,
    captionColor: typeof raw.captionColor === "string" ? normalizeHex(raw.captionColor) : null,
    showLogo: raw.showLogo !== false,
  };
}

function promisify<T>(req: IDBRequest<T>): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    req.onsuccess = () => resolve(req.result);
    req.onerror = () => reject(req.error ?? new Error("IndexedDB request failed"));
  });
}

/** Open (and create) the store, or null when IndexedDB can't serve us. */
function openDb(): Promise<IDBDatabase | null> {
  if (!hasIndexedDB()) return Promise.resolve(null);
  return new Promise<IDBDatabase | null>((resolve) => {
    let settled = false;
    const done = (db: IDBDatabase | null) => {
      if (settled) return;
      settled = true;
      resolve(db);
    };
    const timer = setTimeout(() => done(null), IDB_TIMEOUT_MS);
    const finish = (db: IDBDatabase | null) => {
      clearTimeout(timer);
      done(db);
    };
    let req: IDBOpenDBRequest;
    try {
      req = indexedDB.open(DB_NAME, DB_VERSION);
    } catch {
      finish(null); // storage disabled entirely (some private modes)
      return;
    }
    req.onupgradeneeded = () => {
      const db = req.result;
      if (!db.objectStoreNames.contains(STORE_NAME)) db.createObjectStore(STORE_NAME);
    };
    req.onsuccess = () => {
      const db = req.result;
      // An older DB from a future/parallel schema without our store is not
      // something to repair mid-session; degrade to memory.
      finish(db.objectStoreNames.contains(STORE_NAME) ? db : null);
    };
    req.onerror = () => finish(null);
    req.onblocked = () => finish(null);
  });
}

async function withStore<T>(
  mode: IDBTransactionMode,
  work: (store: IDBObjectStore) => Promise<T>
): Promise<T | null> {
  const db = await openDb();
  if (!db) return null;
  try {
    const tx = db.transaction(STORE_NAME, mode);
    const result = await work(tx.objectStore(STORE_NAME));
    return result;
  } catch {
    return null;
  } finally {
    try {
      db.close();
    } catch {
      // Closing a already-errored connection is not worth reporting.
    }
  }
}

/** The stored kit. Never throws: EMPTY_KIT on any failure. */
export async function loadKit(): Promise<BrandKit> {
  try {
    const stored = await withStore("readonly", (store) =>
      promisify<unknown>(store.get(RECORD_KEY))
    );
    if (stored === null || stored === undefined) return memoryKit ?? EMPTY_KIT;
    const kit = coerceKit(stored);
    memoryKit = kit;
    return kit;
  } catch {
    return memoryKit ?? EMPTY_KIT;
  }
}

/** Persist the kit. Never throws; a failed write still applies this session. */
export async function saveKit(kit: BrandKit): Promise<void> {
  const clean = coerceKit(kit);
  memoryKit = clean;
  try {
    await withStore("readwrite", (store) => promisify(store.put(clean, RECORD_KEY)));
  } catch {
    // Quota, private mode, a closed connection — the in-memory kit stands.
  }
}

/** Forget the kit everywhere. Never throws. */
export async function clearKit(): Promise<void> {
  memoryKit = null;
  try {
    await withStore("readwrite", (store) => promisify(store.delete(RECORD_KEY)));
  } catch {
    // Nothing to do — the in-memory kit is already gone.
  }
}
